core/src/main/java/org/apache/calcite/adapter/jdbc/JdbcPreparedStatementUtils.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.adapter.jdbc;
+
+import org.apache.calcite.DataContext;
+
+import java.math.BigDecimal;
+import java.net.URL;
+import java.sql.Blob;
+import java.sql.Clob;
+import java.sql.Date;
+import java.sql.NClob;
+import java.sql.PreparedStatement;
+import java.sql.Ref;
+import java.sql.RowId;
+import java.sql.SQLException;
+import java.sql.SQLXML;
+import java.sql.Time;
+import java.sql.Timestamp;
+import java.util.function.Consumer;
+
+/**
+ * Utils for handling prepared statement decoration, like setting dynamic parameters
+ */
+public class JdbcPreparedStatementUtils {
+
+  private JdbcPreparedStatementUtils() {
+  }
+
+  public static Consumer<PreparedStatement> createPreparedStatementConsumer(Integer[] indexes,
+                                                                            DataContext context) {
+    return pstm -> {
+      int dynamicParamIndex = 1;
+      for (int index : indexes) {
+        setDynamicParam(dynamicParamIndex, context.get("?" + index), pstm);
+        dynamicParamIndex++;
+      }
+    };
+
+  }
+
+  private static void setDynamicParam(Integer i, Object value, PreparedStatement pstm) {
+    try {
+      if (value == null) {
+        pstm.setObject(i, null);
```

Comment:
That should never be used.
`null` values should be passed via `.setNull(index, type)` since DB should know the type of the `null`

Reply:
This is because from the current state of SqlString and DataContext we don't now former setXXX method invoked on original statement.

Reply:
Oh, I see.

On the other hand, `DataContextImpl` has perfect type information: https://github.com/apache/calcite/blob/8f14945461b3c20cba3f2d794a6ac950b7687da9/core/src/main/java/org/apache/calcite/jdbc/CalciteConnectionImpl.java#L290

I wonder if we can add `TypedValue.bind(PreparedStatement, index)` kind of method.
How do you like it?

Reply:
I will check this, current binding in the util is more than naive.

Reply:
I've checked that avatica does not support much of the setXXX methods on PreparedStatements (AvaticaSite.setXXX, most are not implemented). Also TypedValue is a Avatica class.

I would phase this out, so this pull (after some refinements) could be merged and under new task change the Avatica and then adapt the Calcite.  With two smaller steps we could have a feature quicker, even not fully supported (albeit supported up to the support offered by Avatica)

AvaticaSite could be modified as following:

`
public void setNCharacterStream(Reader value, long length, Consumer<PreparedStatement> binder) {
  lots[index] = wrap(ColumnMetaData.Rep.OBJECT, o, binder);
}
` 

then AvaticaPreparedStatement

`
   public void setNCharacterStream(int parameterIndex, Reader value) throws SQLException {
        this.getSite(parameterIndex).setNCharacterStream(value, pstm -> pstm.setNCharacterStream(parameterIndex, value));
    }
`

Then TypedValue would have full information on the end user intention how to bind params on the statement.